Add GFv1-NHDv2 xwalk

lekoenig opened the pull request:
The code changes here build a crosswalk table between PRMS segments (NHGFv1) and NHDPlusV2 flowlines for the Delaware River Basin (DRB). The xwalk target was initially created as part of the DRB inland salinity project, and the full commit history can be referenced within that project's [repo](https://github.com/USGS-R/drb-inland-salinity-ml). 

Below is a preview of the output table, where `comid_down` represents the NHDPlusV2 COMID at the downstream end of each PRMS segment, `comid_seg` represents the NHDPlusV2 COMID's that intersect each PRMS segment, and `comid_cat` represents all of the NHDPlusV2 COMID's that directly drain to the PRMS segment, including headwater tributaries. 
```
> tar_load(p2_prms_nhdv2_xwalk)
> head(p2_prms_nhdv2_xwalk)
# A tibble: 6 x 4
  PRMS_segid comid_down comid_seg                                                       comid_cat                                                                
  <chr>           <int> <chr>                                                           <chr>                                                                    
1 1_1           1748535 1748535                                                         1748535                                                                  
2 10_1          1748727 1748727                                                         1748727                                                                  
3 11_1          1748743 1748743                                                         1748743                                                                  
4 111_1         2612794 2612794                                                         2612794                                                                  
5 112_1         2612822 2612792;2612808;2612818;2612820;2612822;2612922;2612924;2612928 2612788;2612790;2612792;2612796;2612798;2612800;2612802;2612804;2612806;~
6 113_1         2612826 2612826                                                         2612824;2612826;2612960                                                  
> 
```

This PR also includes two log files that get committed. These log files contain the data hash for each target and the sf external dependency versions that were used to build the xwalk table.

I'm open to any suggestions you have for documenting our steps. Thanks!

jds485 commented on 2_process.R:
```diff
@@ -0,0 +1,48 @@
+  ),
+  
+  # Create and save indicator file
+  tar_target(
+    p2_data_summary_csv,
+    write_ind_files("2_process/log/GFv1_data_summary.csv",
+                    target_names = c("p1_GFv1_reaches_sf","p1_GFv1_catchments_sf",
+                                     "p1_nhdv2reaches_sf","p2_prms_nhdv2_xwalk","p2_prms_nhdv2_xwalk_rds")),
+    format = "file"
+  ),
```
I'm running the pipeline for the first time and I noticed that this target built first. I'll check to see if it builds again after the targets in `target_names` are built, or if it appears as outdated after the pipeline completes.

jds485 replied:
I can confirm that this file is blank and the target does not appear as outdated

jds485 replied:
well, it has headers:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jsmith/AppData/Local/Temp/4/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jsmith/AppData/Local/Temp/4/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->

</head>

<body link="#0563C1" vlink="#954F72">



tar_name | hash
-- | --




</body>

</html>

jds485 replied:
Oh, this may have happened because your log files are in the repo, but the target objects are not. So, when I ran `tar_make`, the first detected change was that all the targets are now missing and it updated the log file. Definitely not something we accounted for with these files!

jds485 replied:
It could be that the `target_names` need to somehow be specified unquoted to be tracked by targets. `tar_visnetwork` doesn't show these `target_names` as dependencies to the log file.

lekoenig replied:
Great catch! I think this was happening because the log files are in the repo before the targets get built (your point above). I added a cue to the log file targets so that they get built every time you run `tar_make()`. I think this solves the problem, but let me know if you see otherwise on your end.

lekoenig replied:
> It could be that the target_names need to somehow be specified unquoted to be tracked by targets.

The character string gets passed to a function we've defined to subset the `tar_meta()` table, so I don't think that's the crux of the issue here.

jds485 replied:
I tried rebuilding from scratch and the `p2_data_summary_csv` target is still built first, resulting in a file with only headers. Looking at the target invalidation rules [here](https://rdrr.io/cran/targets/man/tar_cue.html), I wonder if committing the meta file and the object for `p2_data_summary_csv` would solve the problem. That way `p2_data_summary_csv` would only update if a dependency changed.

> The character string gets passed to a function

I was wondering if this properly tracks target dependencies when the target is only specified as a string. The dependency diagram does not show the targets contributing to `p2_data_summary_csv`. Looking into what [dependencies are tracked](https://books.ropensci.org/targets/targets.html#dependencies), I don't think strings would be tracked

```
Not all of the objects from tar_deps() actually register as dependencies. When it comes to detecting dependencies, targets only recognizes

Other targets (such as raw_data_file).
Functions and objects in the main environment. This environment is almost always the global environment of the R process that runs the target script file (default: _targets.R) so these dependencies are usually going to be the custom functions and objects you write yourself.
```

jds485 replied:
I tried with `deparse(substitute(p1_GFv1_reaches_sf))` but that also didn't track the target dependency

lekoenig replied:
The problem here is that the target `p2_data_summary_csv` is building before any of the other data targets in the pipeline and so for fresh builds the call to `tar_meta()` was coming up blank and returning an empty data table where hashes should be. 

I'm not sure that this issue stems from the fact that the log files get committed since changing `tar_cue` to `always` had no effect on this behavior. Rather, as you point out above @jds485, I think we need to impose dependencies on the data summary target. This is hacky - so I'm interested in cleaner fixes - but for now I've opted to add an argument called `force_dep` to the function that creates the data summary table. As the name implies, `force_dep` just forces the dependency with an upstream target so that the targets get built in our desired order (i.e. build the metadata table _after_ building the crosswalk table). 

```
  # Create and save indicator file
  # argument force_dep must contain name of an upstream target to force dependencies
  # and build this target when a log file already exists
  tar_target(
    p2_data_summary_csv,
    write_ind_files("2_process/log/GFv1_data_summary.csv",
                    force_dep = p2_prms_nhdv2_xwalk_csv,
                    target_names = c("p1_GFv1_reaches_sf","p1_GFv1_catchments_sf","p1_nhdv2reaches_sf",
                                     "p2_prms_nhdv2_xwalk","p2_prms_nhdv2_xwalk_csv")),
    format = "file"),
```

jds485 commented on 2_process/src/write_ind_files.R:
```diff
@@ -0,0 +1,41 @@
+}
+
+
+write_session_info <- function(fileout){
```
I'm not sure if R version and OS matter, but could add the output of `Sys.info()` and `R.version()` here, too.

lekoenig replied:
I added the R version and the user to the log file.

jds485 commented:
I think your PR comments could be added to the README file

jds485 commented on .gitignore:
```diff
@@ -0,0 +1,23 @@
+_targets/
+
+#All output
+*/out/*
```
Are you planning to save the crosswalk table to an output folder in the repo?

jds485 replied:
Could add !2_process/out/GFv1_NHDv2_xwalk.rds to not ignore the table

lekoenig replied:
I wasn't originally planning on committing the xwalk table but have opted to in this revision.

jds485 commented on 2_process.R:
```diff
@@ -0,0 +1,48 @@
+  ),
+  
+  # Save GFv1-NHDv2 xwalk table 
+  tar_target(
```
Could you save a csv or txt target, too? I could see someone wanting to check a single reach quickly by looking at the table

lekoenig replied:
I ended up saving the xwalk table as a csv instead of rds. This has the advantages of being human readable and is more language agnostic.

jds485 reviewed (changes requested):
See change request comments on the log file and .gitignore, and a suggestion to add your PR comments to the README for the repo

lekoenig commented:
Thanks, @jds485, I appreciate your help with this review. I think I've addressed your comments, but I'm interested to know whether my changes fix the log file issue you pointed out.

jds485 commented on README.md:
```diff
@@ -1,2 +1,12 @@
-Code repo to prepare the river network for modeling in the Delaware River Basin
+Code repo to prepare the river network for modeling in the Delaware River Basin  
+  
+## Prep the DRB river network for modeling  
+Nothing here yet. Functions to prep the DRB network are currently included in the delaware-model-prep [repo](https://github.com/USGS-R/delaware-model-prep).  
```
Are you planning to add more to this repo?

lekoenig replied:
Yeah, possibly. The delaware-model-prep repo is pretty bulky, so we've talked about moving the network prep steps outside of that repo if/when we shift from GFv1 to GFv2.

jds485 commented:
Changes look good! Let me try building the repo from scratch and see if the log files build correctly.

jds485 reviewed (approved):
Hacky log file dependency method works! I think this is ready to merge.
